### PR TITLE
Update Custom Callback Docs

### LINKED ISSUE #17149: Multiple callback instances
### 📚 Documentation

```
class Counter(Callback):
    def __init__(self, what="epochs", verbose=True):
        self.what = what
        self.verbose = verbose
        self.state = {"epochs": 0, "batches": 0}

    @property
    def state_key(self):
        # note: we do not include `verbose` here on purpose
        return self._generate_state_key(what=self.what)

    def on_train_epoch_end(self, *args, **kwargs):
        if self.what == "epochs":
            self.state["epochs"] += 1

    def on_train_batch_end(self, *args, **kwargs):
        if self.what == "batches":
            self.state["batches"] += 1

    def load_state_dict(self, state_dict):
        self.state.update(state_dict)

    def state_dict(self):
        return self.state.copy()


# two callbacks of the same type are being used
trainer = Trainer(callbacks=[Counter(what="epochs"), Counter(what="batches")])
```

The code above is quoted directly from https://lightning.ai/docs/pytorch/stable/extensions/callbacks.html as an example to support multiple instances of the callback to the `Trainer`. 

In the example, `self._generate_state_key(what=self.what)` is used, which is a protected method from the parent `Callback` class. I am just wondering if it would be better to provide a public method in the parent class to expose that functionality? Thanks.

cc @borda

### PR DESCRIPTION
**Pull request by yhl48**
## What does this PR do?

Fixes #17149

@awaelchli 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by yhl48** [comment]
There is a problem building docs, locally it is the problem raised in #16158

File: docs/source-pytorch/extensions/callbacks_state.rst [hunk 1]
```diff
@@ -26,6 +26,9 @@ the following example.
             # note: we do not include `verbose` here on purpose
             return self._generate_state_key(what=self.what)
 
+        def _generate_state_key(self, **kwargs: Any) -> str:
+            return f"{self.__class__.__qualname__}{repr(kwargs)}"
```
Comment (awaelchli) [inline docs/source-pytorch/extensions/callbacks_state.rst:30]:
```suggestion
            return f"Counter[what={self.what}]"
```
I think this is more readable for the example

Comment (yhl48) [reply docs/source-pytorch/extensions/callbacks_state.rst]: What about this?

```
@property
def state_key(self):
      return f"Counter[what={self.what}]"
```

Comment (awaelchli) [reply docs/source-pytorch/extensions/callbacks_state.rst]: Yeah that's what I meant 😃 I would avoid the _generate_state_key method completely.

**Review by carmocca — APPROVED** [review]

File: docs/source-pytorch/extensions/callbacks_state.rst [hunk 2]
```diff
@@ -23,8 +23,7 @@ the following example.
 
         @property
         def state_key(self):
-            # note: we do not include `verbose` here on purpose
```
Comment (awaelchli) [inline docs/source-pytorch/extensions/callbacks_state.rst:26]: the note should stay, it is actually relevant in this example :)

**Review by awaelchli — APPROVED** [review]

File: docs/source-pytorch/extensions/callbacks_state.rst [hunk 3]
```diff
@@ -24,7 +24,7 @@ the following example.
         @property
         def state_key(self):
```
Comment (Borda) [inline docs/source-pytorch/extensions/callbacks_state.rst:25]:
```suggestion
        def state_key(self) -> str:
```

**Review by Borda — APPROVED** [review]
